File: rust/src/indicators/welles_wilder/relative_strength_index/relative_strength_index.rs
use crate::entities::bar::Bar;
use crate::entities::bar_component::{component_value as bar_component_value, BarComponent, DEFAULT_BAR_COMPONENT};
use crate::entities::quote::Quote;
use crate::entities::quote_component::{component_value as quote_component_value, QuoteComponent, DEFAULT_QUOTE_COMPONENT};
use crate::entities::scalar::Scalar;
use crate::entities::trade::Trade;
use crate::entities::trade_component::{component_value as trade_component_value, TradeComponent, DEFAULT_TRADE_COMPONENT};
use crate::indicators::core::build_metadata::{build_metadata, OutputText};
use crate::indicators::core::component_triple_mnemonic::component_triple_mnemonic;
use crate::indicators::core::identifier::Identifier;
use crate::indicators::core::indicator::{Indicator, Output};
use crate::indicators::core::line_indicator::LineIndicator;
use crate::indicators::core::metadata::Metadata;

// ---------------------------------------------------------------------------
// Params
// ---------------------------------------------------------------------------

/// Parameters for the Relative Strength Index indicator.
pub struct RelativeStrengthIndexParams {
    /// Number of periods. Must be >= 2. Default is 14.
    pub length: usize,
    pub bar_component: Option<BarComponent>,
    pub quote_component: Option<QuoteComponent>,
    pub trade_component: Option<TradeComponent>,
}

impl Default for RelativeStrengthIndexParams {
    fn default() -> Self {
        Self {
            length: 14,
            bar_component: None,
            quote_component: None,
            trade_component: None,
        }
    }
}

// ---------------------------------------------------------------------------
// Output
// ---------------------------------------------------------------------------

/// Enumerates the outputs of the RSI indicator.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[repr(u8)]
pub enum RelativeStrengthIndexOutput {
    /// The scalar value of the RSI.
    Value = 1,
}

// ---------------------------------------------------------------------------
// Indicator
// ---------------------------------------------------------------------------

const EPSILON: f64 = 1e-8;

/// Welles Wilder's Relative Strength Index (RSI).
///
/// RSI measures the magnitude of recent price changes to evaluate overbought
/// or oversold conditions. It oscillates between 0 and 100.
pub struct RelativeStrengthIndex {
    line: LineIndicator,
    length: usize,
    count: i32,
    previous_sample: f64,
    previous_gain: f64,
    previous_loss: f64,
    value: f64,
    primed: bool,
}

impl RelativeStrengthIndex {
    /// Creates a new RelativeStrengthIndex from the given parameters.
    pub fn new(params: &RelativeStrengthIndexParams) -> Result<Self, String> {
        if params.length < 2 {
            return Err("invalid relative strength index parameters: length should be greater than 1".to_string());
        }

        let bc = params.bar_component.unwrap_or(DEFAULT_BAR_COMPONENT);
        let qc = params.quote_component.unwrap_or(DEFAULT_QUOTE_COMPONENT);
        let tc = params.trade_component.unwrap_or(DEFAULT_TRADE_COMPONENT);

        let bar_func = bar_component_value(bc);
        let quote_func = quote_component_value(qc);
        let trade_func = trade_component_value(tc);

        let mnemonic = format!("rsi({}{})", params.length, component_triple_mnemonic(bc, qc, tc));
        let description = format!("Relative Strength Index {}", mnemonic);

        let line = LineIndicator::new(mnemonic, description, bar_func, quote_func, trade_func);

        Ok(Self {
            line,
            length: params.length,
            count: -1,
            previous_sample: 0.0,
            previous_gain: 0.0,
            previous_loss: 0.0,
            value: f64::NAN,
            primed: false,
        })
    }

    /// Core update with a single sample value.
    pub fn update(&mut self, sample: f64) -> f64 {
        if sample.is_nan() {
            return sample;
        }

        self.count += 1;

        if self.count == 0 {
            self.previous_sample = sample;
            return self.value;
        }

        let temp = sample - self.previous_sample;
        self.previous_sample = sample;

        if !self.primed {
            // Accumulation phase: count 1..length-1.
            if temp < 0.0 {
                self.previous_loss -= temp;
            } else {
                self.previous_gain += temp;
            }

            if (self.count as usize) < self.length {
                return self.value;
            }

            // Priming: count == length.
            self.previous_gain /= self.length as f64;
            self.previous_loss /= self.length as f64;
            self.primed = true;
        } else {
            // Wilder's smoothing.
            self.previous_gain *= (self.length - 1) as f64;
            self.previous_loss *= (self.length - 1) as f64;

            if temp < 0.0 {
                self.previous_loss -= temp;
            } else {
                self.previous_gain += temp;
            }

            self.previous_gain /= self.length as f64;
            self.previous_loss /= self.length as f64;
        }

        let sum = self.previous_gain + self.previous_loss;
        if sum > EPSILON {
            self.value = 100.0 * self.previous_gain / sum;
        } else {
            self.value = 0.0;
        }

        self.value
    }
}

impl Indicator for RelativeStrengthIndex {
    fn is_primed(&self) -> bool {
        self.primed
    }

    fn metadata(&self) -> Metadata {
        build_metadata(
            Identifier::RelativeStrengthIndex,
            &self.line.mnemonic,
            &self.line.description,
            &[OutputText {
                mnemonic: self.line.mnemonic.clone(),
                description: self.line.description.clone(),
            }],
        )
    }

    fn update_scalar(&mut self, sample: &Scalar) -> Output {
        let value = self.update(sample.value);
        vec![Box::new(Scalar::new(sample.time, value))]
    }

    fn update_bar(&mut self, sample: &Bar) -> Output {
        let sample_value = (self.line.bar_func)(sample);
        let value = self.update(sample_value);
        vec![Box::new(Scalar::new(sample.time, value))]
    }

    fn update_quote(&mut self, sample: &Quote) -> Output {
        let sample_value = (self.line.quote_func)(sample);
        let value = self.update(sample_value);
        vec![Box::new(Scalar::new(sample.time, value))]
    }

    fn update_trade(&mut self, sample: &Trade) -> Output {
        let sample_value = (self.line.trade_func)(sample);
        let value = self.update(sample_value);
        vec![Box::new(Scalar::new(sample.time, value))]
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn test_input_1() -> Vec<f64> {
        vec![
            91.15, 90.50, 92.55, 94.70, 95.55, 94.00, 91.30, 91.95, 92.45, 93.80,
            92.50, 94.55, 96.75, 97.80, 98.40, 98.15, 96.70, 98.85, 98.90, 100.50,
            102.60, 104.80, 103.80, 103.10, 102.00,
        ]
    }

    fn test_expected_1() -> Vec<f64> {
        vec![
            f64::NAN, f64::NAN, f64::NAN, f64::NAN, f64::NAN,
            f64::NAN, f64::NAN, f64::NAN, f64::NAN,
            60.6425702811244, 54.2677448337826, 61.4558190165176, 67.6034767388667,
            70.1590191481383, 71.5992400904851, 70.0152589447766, 61.1833361324987,
            67.9312249318593, 68.076417836971, 72.5504646296262, 77.2568847385616,
            81.0801123570899, 74.6619680507228, 70.2808713845906, 63.6754215506388,
        ]
    }

    fn test_input_2() -> Vec<f64> {
        vec![
            44.34, 44.09, 43.61, 44.33, 44.83, 45.10, 45.42, 45.84, 46.08, 45.89,
            46.03, 45.61, 46.28, 46.28, 46.00, 46.03, 46.41, 46.22, 45.64, 46.21,
            46.25, 45.71, 46.45, 45.78, 45.35, 44.03, 44.18, 44.22, 44.57, 43.42,
            42.66, 43.13, 44.94, 43.61, 44.33, 44.83, 45.10, 45.42, 45.84, 46.08,
            45.89, 46.03, 45.61, 46.28, 46.28, 46.00, 46.03, 46.41, 46.22, 45.64,
            46.21, 46.25, 45.71, 46.45, 45.78, 45.35, 44.03, 44.18, 44.22, 44.57,
            43.42, 42.66, 43.13, 44.94, 43.61, 44.33, 44.83, 45.10, 45.42, 45.84,
            46.08, 45.89, 46.03, 45.61, 46.28, 46.28, 46.00, 46.03, 46.41, 46.22,
            45.64, 46.21, 46.25, 45.71, 46.45, 45.78, 45.35, 44.03, 44.18, 44.22,
            44.57, 43.42, 42.66, 43.13, 44.94, 43.61, 44.33, 44.83, 45.10, 45.42,
            45.84, 46.08, 45.89, 46.03, 45.61, 46.28, 46.28, 46.00, 46.03, 46.41,
            46.22, 45.64, 46.21, 46.25, 45.71, 46.45, 45.78, 45.35, 44.03, 44.18,
            44.22, 44.57, 43.42, 42.66, 43.13, 44.94, 43.61, 44.33, 44.83, 45.10,
            45.42, 45.84, 46.08, 45.89, 46.03, 45.61, 46.28, 46.28, 46.00, 46.03,
            46.41, 46.22, 45.64, 46.21, 46.25, 45.71, 46.45, 45.78, 45.35, 44.03,
            44.18, 44.22, 44.57, 43.42, 42.66, 43.13, 44.94, 43.61, 44.33, 44.83,
            45.10, 45.42, 45.84, 46.08, 45.89, 46.03, 45.61, 46.28, 46.28, 46.00,
            46.03, 46.41, 46.22, 45.64, 46.21, 46.25, 45.71, 46.45, 45.78, 45.35,
            44.03, 44.18, 44.22, 44.57, 43.42, 42.66, 43.13, 44.94, 43.61, 44.33,
            44.83, 45.10, 45.42, 45.84, 46.08, 45.89, 46.03, 45.61, 46.28, 46.28,
            46.00, 46.03, 46.41, 46.22, 45.64, 46.21, 46.25, 45.71, 46.45, 45.78,
            45.35, 44.03, 44.18, 44.22, 44.57, 43.42, 42.66, 43.13, 44.94, 43.61,
            44.33, 44.83, 45.10, 45.42, 45.84, 46.08, 45.89, 46.03, 45.61, 46.28,
            46.28, 46.00, 46.03, 46.41, 46.22, 45.64, 46.21, 46.25, 45.71, 46.45,
            45.78, 45.35, 44.03, 44.18, 44.22, 44.57, 43.42, 42.66, 43.13, 44.94,
            43.61, 44.33,
        ]
    }

    #[test]
    fn test_rsi_update() {
        let tolerance = 1e-9;
        let input = test_input_1();
        let expected = test_expected_1();
        let params = RelativeStrengthIndexParams { length: 9, ..Default::default() };
        let mut rsi = RelativeStrengthIndex::new(&params).unwrap();

        for i in 0..input.len() {
            let act = rsi.update(input[i]);

            if i < 9 {
                assert!(act.is_nan(), "[{}] expected NaN, got {}", i, act);
                continue;
            }

            assert!(
                (act - expected[i]).abs() < tolerance,
                "[{}] expected {}, got {}",
                i,
                expected[i],
                act
            );
        }

        assert!(rsi.update(f64::NAN).is_nan());
    }

    #[test]
    fn test_rsi_update_2() {
        let tolerance = 0.5;
        let input = test_input_2();
        let params = RelativeStrengthIndexParams { length: 14, ..Default::default() };
        let mut rsi = RelativeStrengthIndex::new(&params).unwrap();

        let mut act = f64::NAN;
        for i in 0..input.len() {
            act = rsi.update(input[i]);
            if i < 14 {
                assert!(act.is_nan(), "[{}] expected NaN, got {}", i, act);
            }
        }

        assert!(act >= 0.0 && act <= 100.0, "final RSI should be in [0,100], got {}", act);
    }

    #[test]
    fn test_rsi_is_primed() {
        let params = RelativeStrengthIndexParams { length: 5, ..Default::default() };
        let mut rsi = RelativeStrengthIndex::new(&params).unwrap();

        assert!(!rsi.is_primed());

        for i in 1..=5 {
            rsi.update(i as f64);
            assert!(!rsi.is_primed(), "[{}] should not be primed", i);
        }

        rsi.update(6.0);
        assert!(rsi.is_primed());

        for i in 7..=11 {
            rsi.update(i as f64);
            assert!(rsi.is_primed(), "[{}] should be primed", i);
        }
    }

    #[test]
    fn test_rsi_metadata() {
        let params = RelativeStrengthIndexParams { length: 9, ..Default::default() };
        let rsi = RelativeStrengthIndex::new(&params).unwrap();
        let meta = rsi.metadata();

        assert_eq!(meta.identifier, Identifier::RelativeStrengthIndex);
        assert_eq!(meta.outputs.len(), 1);
        assert_eq!(meta.outputs[0].mnemonic, "rsi(9)");
        assert_eq!(meta.outputs[0].description, "Relative Strength Index rsi(9)");
    }

    #[test]
    fn test_rsi_constructor_validation() {
        let params = RelativeStrengthIndexParams { length: 1, ..Default::default() };
        assert!(RelativeStrengthIndex::new(&params).is_err());

        let params = RelativeStrengthIndexParams { length: 14, ..Default::default() };
        let rsi = RelativeStrengthIndex::new(&params).unwrap();
        assert_eq!(rsi.line.mnemonic, "rsi(14)");
    }

    #[test]
    fn test_rsi_bar_component_mnemonic() {
        let params = RelativeStrengthIndexParams {
            length: 14,
            bar_component: Some(BarComponent::Open),
            ..Default::default()
        };
        let rsi = RelativeStrengthIndex::new(&params).unwrap();
        assert_eq!(rsi.line.mnemonic, "rsi(14, o)");
    }
}

File: zig/src/indicators/welles_wilder/relative_strength_index/relative_strength_index.zig
const std = @import("std");
const math = std.math;


const entities = @import("entities");
const Bar = entities.Bar;
const Quote = entities.Quote;
const Trade = entities.Trade;
const Scalar = entities.Scalar;
const bar_component = entities.bar_component;
const quote_component = entities.quote_component;
const trade_component = entities.trade_component;
const indicator_mod = @import("../../core/indicator.zig");
const line_indicator_mod = @import("../../core/line_indicator.zig");
const build_metadata_mod = @import("../../core/build_metadata.zig");
const component_triple_mnemonic_mod = @import("../../core/component_triple_mnemonic.zig");
const identifier_mod = @import("../../core/identifier.zig");
const metadata_mod = @import("../../core/metadata.zig");

const OutputArray = indicator_mod.OutputArray;
const LineIndicator = line_indicator_mod.LineIndicator;
const Identifier = identifier_mod.Identifier;
const Metadata = metadata_mod.Metadata;

/// Enumerates the outputs of the Relative Strength Index indicator.
pub const RelativeStrengthIndexOutput = enum(u8) {
    /// The RSI value (0..100).
    value = 1,
};

/// Parameters for the Relative Strength Index indicator.
pub const RelativeStrengthIndexParams = struct {
    /// The smoothing length. Must be >= 2. Default is 14.
    length: usize = 14,
    /// Bar component to extract. `null` means use default (Close).
    bar_component: ?bar_component.BarComponent = null,
    /// Quote component to extract. `null` means use default (Mid).
    quote_component: ?quote_component.QuoteComponent = null,
    /// Trade component to extract. `null` means use default (Price).
    trade_component: ?trade_component.TradeComponent = null,
};

/// Welles Wilder's Relative Strength Index (RSI).
///
/// RSI = 100 * avg_gain / (avg_gain + avg_loss)
///
/// Uses Wilder's smoothing:
///   avg_gain(n) = avg_gain(n-1) * (n-1)/n + gain_today / n
///   avg_loss(n) = avg_loss(n-1) * (n-1)/n + loss_today / n
///
/// The indicator is not primed until `length` samples have been processed.
pub const RelativeStrengthIndex = struct {
    line: LineIndicator,
    length: usize,
    count: isize,
    previous_sample: f64,
    previous_gain: f64,
    previous_loss: f64,
    value: f64,
    primed: bool,
    mnemonic_buf: [64]u8,
    mnemonic_len: usize,
    description_buf: [128]u8,
    description_len: usize,

    pub const Error = error{ InvalidLength, MnemonicTooLong, DescriptionTooLong };

    pub fn init(params: RelativeStrengthIndexParams) Error!RelativeStrengthIndex {
        if (params.length < 2) return error.InvalidLength;

        const bc = params.bar_component orelse bar_component.default_bar_component;
        const qc = params.quote_component orelse quote_component.default_quote_component;
        const tc = params.trade_component orelse trade_component.default_trade_component;

        var triple_buf: [64]u8 = undefined;
        const triple = component_triple_mnemonic_mod.componentTripleMnemonic(&triple_buf, bc, qc, tc);

        var mnemonic_buf: [64]u8 = undefined;
        const mnemonic_slice = std.fmt.bufPrint(&mnemonic_buf, "rsi({d}{s})", .{ params.length, triple }) catch return error.MnemonicTooLong;
        const mnemonic_len = mnemonic_slice.len;

        var description_buf: [128]u8 = undefined;
        const desc_slice = std.fmt.bufPrint(&description_buf, "Relative Strength Index {s}", .{mnemonic_slice}) catch return error.DescriptionTooLong;
        const description_len = desc_slice.len;

        return .{
            .line = LineIndicator.new(mnemonic_slice, desc_slice, bc, qc, tc),
            .length = params.length,
            .count = -1,
            .previous_sample = 0,
            .previous_gain = 0,
            .previous_loss = 0,
            .value = math.nan(f64),
            .primed = false,
            .mnemonic_buf = mnemonic_buf,
            .mnemonic_len = mnemonic_len,
            .description_buf = description_buf,
            .description_len = description_len,
        };
    }

    pub fn deinit(_: *RelativeStrengthIndex) void {}
    pub fn fixSlices(self: *RelativeStrengthIndex) void {
        self.line.mnemonic = self.mnemonic_buf[0..self.mnemonic_len];
        self.line.description = self.description_buf[0..self.description_len];
    }

    /// Core update given a single sample value.
    pub fn update(self: *RelativeStrengthIndex, sample: f64) f64 {
        if (math.isNan(sample)) return math.nan(f64);

        self.count += 1;

        if (self.count == 0) {
            self.previous_sample = sample;
            return self.value;
        }

        const temp = sample - self.previous_sample;
        self.previous_sample = sample;

        if (!self.primed) {
            if (temp < 0) {
                self.previous_loss -= temp;
            } else {
                self.previous_gain += temp;
            }

            if (self.count < self.length) {
                return self.value;
            }

            self.previous_gain /= @floatFromInt(self.length);
            self.previous_loss /= @floatFromInt(self.length);
            self.primed = true;
        } else {
            const n: f64 = @floatFromInt(self.length);
            self.previous_gain *= n - 1;
            self.previous_loss *= n - 1;

            if (temp < 0) {
                self.previous_loss -= temp;
            } else {
                self.previous_gain += temp;
            }

            self.previous_gain /= n;
            self.previous_loss /= n;
        }

        const sum = self.previous_gain + self.previous_loss;
        if (sum > 1e-8) {
            self.value = 100.0 * self.previous_gain / sum;
        } else {
            self.value = 0.0;
        }

        return self.value;
    }

    /// Update using a single sample value (convenience method).
    pub fn updateSample(self: *RelativeStrengthIndex, sample: f64) f64 {
        return self.update(sample);
    }

    pub fn isPrimed(self: *const RelativeStrengthIndex) bool {
        return self.primed;
    }

    pub fn getMetadata(self: *const RelativeStrengthIndex, out: *Metadata) void {
        build_metadata_mod.buildMetadata(out, Identifier.relative_strength_index, self.line.mnemonic, self.line.description, &.{
            .{ .mnemonic = self.line.mnemonic, .description = self.line.description },
        });
    }

    fn makeOutput(self: *const RelativeStrengthIndex, time: i64) OutputArray {
        var out = OutputArray{};
        out.append(.{ .scalar = Scalar{ .time = time, .value = self.value } });
        return out;
    }

    pub fn updateScalar(self: *RelativeStrengthIndex, sample: *const Scalar) OutputArray {
        _ = self.update(sample.value);
        return self.makeOutput(sample.time);
    }

    pub fn updateBar(self: *RelativeStrengthIndex, sample: *const Bar) OutputArray {
        const v = self.line.extractBar(sample);
        _ = self.update(v);
        return self.makeOutput(sample.time);
    }

    pub fn updateQuote(self: *RelativeStrengthIndex, sample: *const Quote) OutputArray {
        const v = self.line.extractQuote(sample);
        _ = self.update(v);
        return self.makeOutput(sample.time);
    }

    pub fn updateTrade(self: *RelativeStrengthIndex, sample: *const Trade) OutputArray {
        const v = self.line.extractTrade(sample);
        _ = self.update(v);
        return self.makeOutput(sample.time);
    }

    // --- Indicator interface ---

    pub fn indicator(self: *RelativeStrengthIndex) indicator_mod.Indicator {
        return indicator_mod.Indicator{
            .ptr = @ptrCast(self),
            .vtable = &.{
                .isPrimed = vtableIsPrimed,
                .metadata = vtableMetadata,
                .updateScalar = vtableUpdateScalar,
                .updateBar = vtableUpdateBar,
                .updateQuote = vtableUpdateQuote,
                .updateTrade = vtableUpdateTrade,
            },
        };
    }

    fn vtableIsPrimed(ptr: *const anyopaque) bool {
        const self: *const RelativeStrengthIndex = @ptrCast(@alignCast(ptr));
        return self.isPrimed();
    }

    fn vtableMetadata(ptr: *const anyopaque, out: *Metadata) void {
        const self: *const RelativeStrengthIndex = @ptrCast(@alignCast(ptr));
        self.getMetadata(out);
    }

    fn vtableUpdateScalar(ptr: *anyopaque, sample: *const Scalar) OutputArray {
        const self: *RelativeStrengthIndex = @ptrCast(@alignCast(ptr));
        return self.updateScalar(sample);
    }

    fn vtableUpdateBar(ptr: *anyopaque, sample: *const Bar) OutputArray {
        const self: *RelativeStrengthIndex = @ptrCast(@alignCast(ptr));
        return self.updateBar(sample);
    }

    fn vtableUpdateQuote(ptr: *anyopaque, sample: *const Quote) OutputArray {
        const self: *RelativeStrengthIndex = @ptrCast(@alignCast(ptr));
        return self.updateQuote(sample);
    }

    fn vtableUpdateTrade(ptr: *anyopaque, sample: *const Trade) OutputArray {
        const self: *RelativeStrengthIndex = @ptrCast(@alignCast(ptr));
        return self.updateTrade(sample);
    }
};

// ============================================================================
// Tests
// ============================================================================

const testing = std.testing;
const testdata = @import("testdata.zig");


fn almostEqual(a: f64, b: f64, tolerance: f64) bool {
    if (math.isNan(a) and math.isNan(b)) return true;
    if (math.isNan(a) or math.isNan(b)) return false;
    return @abs(a - b) <= tolerance;
}

// Test data from TA-Lib reference (length=9, 25 entries).
// Test data from TA-Lib reference (length=14, 252 entries, same as DM+).
test "RelativeStrengthIndex update length=9" {
    const tolerance = 1e-9;
    var rsi: RelativeStrengthIndex = try RelativeStrengthIndex.init(.{ .length = 9 });
    rsi.fixSlices();

    for (0..testdata.test_input_1.len) |i| {
        const act = rsi.update(testdata.test_input_1[i]);
        const exp = testdata.test_expected_1[i];

        if (math.isNan(exp)) {
            try testing.expect(math.isNan(act));
        } else {
            try testing.expect(!math.isNan(act));
            try testing.expect(almostEqual(act, exp, tolerance));
        }
    }
}

test "RelativeStrengthIndex update length=14" {
    // Matches Go's TestRelativeStrengthIndexUpdate2: check NaN during warmup,
    // then spot-check final value is in valid RSI range [0, 100].
    var rsi: RelativeStrengthIndex = try RelativeStrengthIndex.init(.{ .length = 14 });
    rsi.fixSlices();

    var act: f64 = math.nan(f64);
    for (0..testdata.test_input_2.len) |i| {
        act = rsi.update(testdata.test_input_2[i]);
        if (i < 14) {
            try testing.expect(math.isNan(act));
        }
    }

    // Final value should be in valid RSI range.
    try testing.expect(act >= 0.0 and act <= 100.0);
}

test "RelativeStrengthIndex constructor validation" {
    const result = RelativeStrengthIndex.init(.{ .length = 1 });
    try testing.expectError(error.InvalidLength, result);
}

test "RelativeStrengthIndex isPrimed length=9" {
    var rsi: RelativeStrengthIndex = try RelativeStrengthIndex.init(.{ .length = 9 });
    rsi.fixSlices();

    try testing.expect(!rsi.isPrimed());

    for (0..9) |_| {
        _ = rsi.update(100.0);
    }
    try testing.expect(!rsi.isPrimed());

    _ = rsi.update(100.0);
    try testing.expect(rsi.isPrimed());
}

test "RelativeStrengthIndex NaN passthrough" {
    var rsi: RelativeStrengthIndex = try RelativeStrengthIndex.init(.{ .length = 14 });
    rsi.fixSlices();

    try testing.expect(math.isNan(rsi.update(math.nan(f64))));
    try testing.expect(math.isNan(rsi.updateSample(math.nan(f64))));
}

test "RelativeStrengthIndex metadata" {
    var rsi: RelativeStrengthIndex = try RelativeStrengthIndex.init(.{ .length = 14 });
    rsi.fixSlices();
    var meta: Metadata = undefined;
    rsi.getMetadata(&meta);

    try testing.expectEqual(Identifier.relative_strength_index, meta.identifier);
    try testing.expectEqual(@as(usize, 1), meta.outputs_len);
}

test "RelativeStrengthIndex updateBar" {
    var rsi: RelativeStrengthIndex = try RelativeStrengthIndex.init(.{ .length = 9 });
    rsi.fixSlices();

    for (0..8) |i| {
        const bar = Bar{ .time = @intCast(i), .open = 0, .high = 0, .low = 0, .close = testdata.test_input_1[i], .volume = 0 };
        _ = rsi.updateBar(&bar);
    }

    const bar = Bar{ .time = 42, .open = 0, .high = 0, .low = 0, .close = testdata.test_input_1[8], .volume = 0 };
    const out = rsi.updateBar(&bar);
    try testing.expectEqual(@as(usize, 1), out.len);
}
